lib/spacesuit/cors_middleware.ex

```diff
@@ -0,0 +1,166 @@
+defmodule Spacesuit.CorsMiddleware do
+  require Logger
+  use Elixometer
+
+  @http_server            Application.get_env(:spacesuit, :http_server)
+  @cors_config            Application.get_env(:spacesuit, :cors_config)
+  @supported_http_methods [:GET, :POST, :PUT, :PATCH, :DELETE, :HEAD, :OPTIONS]
+
+  @timed(key: "timed.corsMiddleware-handle", units: :millisecond)
+  def execute(req, env) do
+    case {req[:headers]["origin"], req[:method]} do
+      {nil, _} ->
+        # http://www.w3.org/TR/cors/#resource-requests
+        # § 6.1.1
+        # If the Origin header is not present terminate this set of steps.
+        Logger.debug "No origin header on request"
+        {:ok, req, env}
+
+      {origin, method} ->
+        if is_nil(origin) or !is_valid_origin(origin) do
+          # If the value of the Origin header is not a case-sensitive match for
+          # any of the values in list of origins, do not set any additional
+          # headers and terminate this set of steps.
+          if @cors_config[:serve_forbidden_origins] do
+            {:ok, req, env}
+          else
+            handle_invalid_cors_request(req)
+          end
+        else if is_same_origin(origin, req) do
+          # Same-origin request
+          {:ok, req, env}
+        else
+          case method do
+            :OPTIONS ->
+              # preflight request check
+              case req[:headers]["access_control_request_headers"] do
+                nil ->
+                  handle_cors_request(req, env)
+                "" ->
+                  handle_invalid_cors_request(req)
+                _ ->
+                  handle_preflight_cors_request(req, env)
+              end
+            _ ->
+              if is_supported_http_method method do
+                handle_cors_request(req, env)
+              else
+                handle_invalid_cors_request(req)
+              end
+          end
+        end
+      end
+    end
+  end
+
+  @spec handle_cors_request(Req.t, Env.t) :: {atom, Req.t} | {atom, Req.t, Env.t}
+  defp handle_cors_request(req, env) do
+    origin = req[:headers]["origin"]
+    if !is_allowed_origin origin do
+      handle_invalid_cors_request(req)
+    end
+    {:ok, :cowboy_req.set_resp_headers(get_cors_origin_headers(origin), req), env}
```

Comment:
this should probably reply to the client rather than pass the request down the chain